File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n=height.length;
        int leftMax[]=new int[n];
        int rightMax[]=new int[n];
        int lval=0,rval=0,res=0;
        for(int i=0;i<n;i++){
            lval=Math.max(lval,height[i]);
            rval=Math.max(rval,height[n-i-1]);
            leftMax[i]=lval;
            rightMax[n-i-1]=rval;
        }
        for(int i=0;i<n;i++){
            res+=Math.min(leftMax[i],rightMax[i])-height[i];
        }
        return res;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n=height.size();
        int lsum[n],rsum[n];
        int lval=0,rval=0,res=0;
        for(int i=0;i<n;i++){
            lval=max(lval,height[i]);
            rval=max(rval,height[n-i-1]);
            lsum[i]=lval;
            rsum[n-i-1]=rval;
        }
        for(int i=0;i<n;i++){
            res+=min(lsum[i],rsum[i])-height[i];
        }
        return res;
    }
};